src/callbacks_step/save_restart_dg.jl

```diff
@@ -47,6 +47,46 @@ function save_restart_file(u, time, dt, timestep,
     return filename
 end
 
+# Version for serial I/O
+function interpolate_restart_file!(u, file,
+                                   mesh, equations, dg, cache,
+                                   nnodes_file, interpolation_matrix)
+    all_variables = zeros(eltype(u),
+                          (nvariables(equations),
+                           ntuple(_ -> nnodes_file, ndims(mesh))...,
+                           nelements(dg, cache)))
+    for v in eachvariable(equations)
+        all_variables[v, .., :] = read(file["variables_$v"])
+    end
+
+    # Perform interpolation
+    for element in eachelement(dg, cache)
+        u[.., element] = multiply_dimensionwise(interpolation_matrix,
+                                                all_variables[.., element])
+    end
+end
+
+# Version for MPI-parallel I/O
+function interpolate_restart_file!(u, file, slice,
+                                   mesh, equations, dg, cache,
```

Comment:
I did not add the dispatch on the `mesh` here as it should not be necessary.

Reply:
Hm yeah true - I restrict this to DGSEM for now.

Good point - I though I am in safe waters as everything happens on the reference element, but maybe I need something like this?

https://github.com/trixi-framework/Trixi.jl/blob/afdb4864105a65458985f46f6947193704f99db6/src/callbacks_step/amr_dg2d.jl#L108-L116

Reply:
So I tested this with the 2D [`elixir_advection_unstructured_flag`](https://github.com/trixi-framework/Trixi.jl/blob/main/examples/p4est_2d_dgsem/elixir_advection_unstructured_flag.jl) where I increase the base polydeg from 3 to 4 and then restart the subsequent run with polydeg = 3 (This is required since the mesh has polydeg 3).

This is the error report for "uniform" refinement:

```julia
 L2 error:       5.43210258e-04
 Linf error:     7.61325588e-03
 
 L2 error:       4.81669631e-05
 Linf error:     7.53667941e-04
 
 L2 error:       3.98543153e-06
 Linf error:     8.58091296e-05
 
 L2 error:       3.11114766e-07
 Linf error:     6.82719964e-06
```

which is not quite 4th order, but certainly more than 3rd order.

Reply:
But it would be good if @andrewwinters5000 could make a statement here :)

Reply:
Yes, this should be okay to have a lower `polydeg` for the mesh compared to the solver. But your projections will need to take the Jacobians into account. In the mapped coordinates it is in fact the quantity `J*u` that is conserved and not simply `u`. However, on static meshes this fact is hidden and the Jacobian is usually divided off for convenience.